presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -218,7 +218,31 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Expression> contex
         public PlanNode visitGroupId(GroupIdNode node, RewriteContext<Expression> context)
         {
             checkState(!DependencyExtractor.extractUnique(context.get()).contains(node.getGroupIdSymbol()), "groupId symbol cannot be referenced in predicate");
-            return context.defaultRewrite(node, context.get());
+
+            List<Symbol> commonGroupingSymbols = node.getCommonGroupingColumns();
+            java.util.function.Predicate<Expression> pushdownEligiblePredicate = conjunct -> DependencyExtractor.extractAll(conjunct).stream()
+                    .allMatch(commonGroupingSymbols::contains);
+
+            Map<Boolean, List<Expression>> conjuncts = extractConjuncts(context.get()).stream().collect(Collectors.partitioningBy(pushdownEligiblePredicate));
+
+            // Push down conjuncts from the inherited predicate that apply to the common grouping columns, or don't apply to any grouping columns
+            PlanNode rewrittenSource = context.rewrite(node.getSource(),  combineConjuncts(conjuncts.get(true)));
+
+            PlanNode output = node;
+            if (rewrittenSource != node.getSource()) {
+                output = new GroupIdNode(node.getId(),
```

Comment:
You can do `ChildReplacer.replaceChildren(node, ImmutableList.of(rewritenSource))`